### PR TITLE
Review of Todo App

### PR DESCRIPTION
**Pull request by akouthoofd**
I've opened this up on a fork of the code just so I can add comments like it we were a real review. Hope this helps!

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+let submitTaskBtn = document.querySelector('.submitTaskBtn');
+submitTaskBtn.addEventListener('click', DOM.submitTask);
+
+let cancelTaskBtn = document.querySelector('.cancelTaskBtn');
+cancelTaskBtn.addEventListener('click', DOM.cancelTask);
+
+let listedProjects = document.querySelectorAll('.project');
+for(let i =0; i < listedProjects.length; i++) {
+    listedProjects[i].addEventListener('mouseup', () => { DOM.displaySelectedProject(listedProjects[i]) })
+};
```
Comment (akouthoofd) [inline src/index.js:18]: Very much like the style of this entrypoint file, though in an attempt to be thorough I should point out that this semicolon (`;`) is technically not supposed to be here. Being a block of code and not a statement, Javascript will see this as a for-loop followed by an empty statement that will no-op (perform an action that literally does nothing).  The nittiest of nit-picks. Does no harm at all but goes against general convention.

And in the spirit of over-analyzing typos and the like the spacing for the `let i = 0` should be consistent between lines 16 & 21. 😝

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+let newProject = document.querySelector('.newProject');
+newProject.addEventListener('click', DOM.addNewProject);
+
+let cancelProject = document.querySelector('.cancel');
+cancelProject.addEventListener('click', DOM.cancelProjectCreation);
+
+let saveProject = document.querySelector('.save');
+saveProject.addEventListener('click', controller.saveNewProject);
+
```
Comment (akouthoofd) [inline src/index.js:34]: Not stuff I would usually mention in a review, but just FYI the general convention is to leave a single newline at the end of a file. The reason is pretty outdated and really doesn't need to be followed, but even GitHub will place a red mark on files that don't have at least one newline at the bottom. Though in this case you have two lines instead of one here. Doesn't matter just a little coding convention knowledge!

File: src/modules.js [hunk 3]
```diff
@@ -0,0 +1,450 @@
+import { parseISO } from "date-fns";
+
+const dataModel = (() => {
+    let allTasks = {
```
Comment (akouthoofd) [inline src/modules.js]: Looks good, though if I'm not mistaken this could be a `const` instead of a `let`.

A good practice in general is to favor immutability (data that cannot be changed) wherever possible. The specifics here are nuanced, but since you are editing the fields inside `allTasks` and never reassigning it (e.g. `allTasks = {<someNewObject>}`), this can be `const` (immutable). This ensures nothing can set this variable, only update the existing one.

This of course comes down to preference, but I like to start with all variables being `const` and move down to `let ` only if I must.

File: src/modules.js [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    const displayTask = (task) => {
+        let displayArea = document.getElementById('main');
+        let div = document.createElement('div');
+        div.classList.add("divCard");
+        div.setAttribute('data-index', task.index);
+        if (task.strike == "none") {
+            div.style.textDecorationLine = "none";
+        } else {
+            div.style.textDecorationLine = "line-through"
+        };
```
Comment (akouthoofd) [inline src/modules.js]: Ignoring the semicolon that shouldn't be here 😜 this is a perfectly valid way to do this, however I have two alternatives you might want to consider.
1. If you have ever heard of a ternary operator, these is a good use-case for it. Not everyone is onboard with the syntax, but personally I like it for simple things like this. It's a shorthand for assigning a variable with an if-else conditionals. Shorthands like this are often called 'syntactic sugar' if you ever hear that term. It's written like this:
```
variableName = (boolean condition) ? <value used if true> : <value used if false>;
```
so in this case:
```
div.style.textDecorationLine = (task.strike === "none") ? "none" : "line-through";
```
So now instead of 5 lines this whole conditional is a single line.

2. Taking this even further. Note you have `task.strike` equal to `none` sometimes and arbitrarily set to `strike` other times. Instead you could just have this variable set to either `none` or `line-through` and then this conditional disappears now only needing:
```
div.style.textDecorationLine = task.strike;
``` 
Tada! 😄

File: src/modules.js [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+
+    const displayTask = (task) => {
+        let displayArea = document.getElementById('main');
+        let div = document.createElement('div');
+        div.classList.add("divCard");
+        div.setAttribute('data-index', task.index);
+        if (task.strike == "none") {
+            div.style.textDecorationLine = "none";
+        } else {
+            div.style.textDecorationLine = "line-through"
```
Comment (akouthoofd) [inline src/modules.js]: Javascript lets you get away with a lot, but for consistency you should add a semicolon here. Don't worry, they're super easy to miss and I still occasionally forget to add one 😅

File: src/modules.js [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        formList.style.visibility = "hidden";
+        document.getElementById('form').reset();
+    };
+
+    const displayTask = (task) => {
+        let displayArea = document.getElementById('main');
+        let div = document.createElement('div');
+        div.classList.add("divCard");
+        div.setAttribute('data-index', task.index);
+        if (task.strike == "none") {
```
Comment (akouthoofd) [inline src/modules.js]: This is a another only Javascript and literally no other language quirk, but it's recommended you use triple equals (`===`) instead of double equals (`==`) for all comparisons.

Reason being Javascript has this ridiculous behavior called 'type coercion' that attempts to convert datatypes automatically to compare them. So if you do `0 === ""` this will and should be `false`, but if you try `0 == ""` this equates to `true` O.o 
image

File: src/modules.js [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+        controller.createTextStrike(img);
+        img = document.createElement('img');
+        img.classList.add('plus');
+        img.src="imgs/plus.svg";
+        div.appendChild(img);
+        controller.createAddToProject(img);
+        img = document.createElement('img');
+        img.classList.add('edit');
+        img.src="imgs/pencil.svg";
+        div.appendChild(img);
```
Comment (akouthoofd) [inline src/modules.js]: There isn't any issues here, however there is a lot of duplication making it hard to identify the real changes here. I would suggest two things here:
1. Add a newline between every logical block of code. So you set this `img` variable to something new and then do stuff with it four times. Every time you change the `img` variable I recommend you add a newline above the reassignment to help readers, including yourself, understand to system of operations.
2. Creating the img, adding the class, and setting the src are repetitive lines of code here. For both readability and maintainability, I'd recommend moving those to a function. Something like:
```
function generateImage(className, imgLocation) {
    const img = document.createElement("img");
    img.classList.add(className);
    img.src = imgLocation;
    return img;
}
```
Side-note: Going to rabbit-hole here a bit, but let me explain why I recommend this. This specific function has an intentional design. It has no "side-effects". This means that calling this function will not make any changes to anything else. It will simple create what you asked for and return it. You can call it once or a dozen times; it will never effect your DOM. Things complicate very quickly when you add side-effects to your functions like `div.appendChild()` which is why I intentionally did not include that in this function.

Now you can clean up this code which should look something like this:
```
        let img = generateImage("trash", "imgs/trash.svg");
        div.appendChild(img);
        controller.createTrashEventHandler(img);

        img = generateImage("check", "imgs/check.svg");
        div.appendChild(img);
        controller.createTextStrike(img);

        img = generateImage("plus", "imgs/plus.svg");
        div.appendChild(img);
        controller.createAddToProject(img);

        img = generateImage("edit", "imgs/pencil.svg");
        div.appendChild(img);
        controller.editTask(img);
        displayArea.appendChild(div);
```
With this it's clear exactly what your trying to do at a glance and the details a abstracted away.

File: src/modules.js [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+        controller.editTask(img);
+        displayArea.appendChild(div);
+    };
+
+    const displayDefaultList = () => {
+        for(let i=0; i< dataModel.allTasks.tasks.length; i++) {
+            displayTask(dataModel.allTasks.tasks[i])
+        };
```
Comment (akouthoofd) [inline src/modules.js]: This for-loop is good, but there exists a better one for this. `dataModel.allTasks.tasks` is an array. Array's have a special for-loop (again just syntactic sugar) that makes them easier to write. It's called a for-of-loop. Looks like this:
```
        for (const task of dataModel.allTasks.tasks) {
            displayTask(task);
        }
```
Nice right? :P

Also more silly nit-picks :) Line 75 should have a semicolon since it's a statement, line 76 shouldn't have one, and the spacing of the for-loop just makes it that much harder to read.

File: src/modules.js [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+            img.onclick = (e) => {
+                let index = img.parentElement.getAttribute('data-index');
+                DOM.createTask();
+                document.getElementById('title').value = dataModel.allTasks.tasks[index].title;
+                document.getElementById('notes').value = dataModel.allTasks.tasks[index].notes;
+                document.getElementById('dueDate').value = dataModel.allTasks.tasks[index].dueDate;
+                document.getElementById('priority').value = dataModel.allTasks.tasks[index].priority;
+                let main = document.querySelector('#main');
+                main.removeChild(img.parentElement);
+                dataModel.allTasks.tasks.splice(index, 1);
```
Comment (akouthoofd) [inline src/modules.js]: This was the only functional problem I encountered in my testing. If you edit a task, this line immediately removes the task being edited for the task list, but if you click 'cancel' it does not add that task back to the list. As a result the task is simply deleted 😱

File: src/modules.js [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        document.getElementById('projectForm').reset();
+    };
+
+    const addProjectToList = (project) => {
+        let parentDiv = document.querySelector('.projectsDiv');
+        let childDiv = document.createElement('div');
+        childDiv.classList.add('project');
+        childDiv.setAttribute('project-type', project.name);
+        childDiv.innerText= project.name;
+        let img = document.createElement('img');
```
Comment (akouthoofd) [inline src/modules.js]: Just pointing out that if you were to create that `generateImage` function I mentioned above lines 104 to 106 can become: 
```
const img = generateImage("projectTrash", "imgs/trash.svg");
```

Also a bit of a repeat, but strategic use of empty lines helps a lot with readability. Personally I would add empty lines here (above line 104) and one above line 109. Clearly indicates to the reader the different operations taking place.

Unrelated nit-picks: no space before the equals on 103 and `let`s could be `const`s 😄

File: src/modules.js [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+        img.addEventListener('click' , (e) => deleteProject(e, img));
+        childDiv.addEventListener('click', () => displaySelectedProject(childDiv));
+        parentDiv.appendChild(childDiv);
+        cancelProjectCreation();
+    };
+
+    const displaySelectedProject = (listedProject) => {
+        let mainDiv = document.getElementById('main');
+        while(mainDiv.firstChild) {
+            mainDiv.removeChild(mainDiv.firstChild)
```
Comment (akouthoofd) [inline src/modules.js]: This was quite clever, though if you want to remove the context of a div you can just do this 😉 
```
mainDiv.innerHTML = "";
```

File: src/modules.js [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+        nameDisplay.innerText = ""
+        let list = listedProject.getAttribute('project-type');
+        if (list == "allTasks") {
+            nameDisplay.innerText = "All Tasks";
+            displayDefaultList()
+        } else if (list == "thisWeek") {
+            nameDisplay.innerText = "This Week";
+            let tasksObj = JSON.parse(localStorage.getItem('allTasks'));
+            if (tasksObj == null) {
+                return
```
Comment (akouthoofd) [inline src/modules.js]: This is good. If you didn't know this is called a guard-clause. Its main value is asserting conditions that must be met to execute the proceeding code and minimizing nesting (essentially code inside curly braces `{}`). You have the right idea, though this specific code is a bit redundant since you wrapped the proceeding code in an else block.

Things would clean up quite nicely if you were to drop the else block and just have the code since that's functionally the same thing with less nesting.

File: src/modules.js [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+                    let currentDate = new Date();
+                    let diff = (parseISO(tasksObj.tasks[i].dueDate).getTime() - currentDate.getTime()) / 86400000;
+                    if (diff <= 7 && diff >= -1) {
+                        displayTask(tasksObj.tasks[i])
+                    }
+                }
+            }
+        } else if (list == "thisMonth") {
+            nameDisplay.innerText = "This Month";
+            let tasksObj = JSON.parse(localStorage.getItem('allTasks'));
```
Comment (akouthoofd) [inline src/modules.js]: So there's nothing wrong with this, but it does break the Model-View-Controller (MVC) design you were shooting for. Editing the `innerText` of a div is very much a View operation where filtering the data to a specific set of tasks is a Model operation. 

I would argue the effort to separate this logic is not worth it given the approach taken, but thought it should be pointed out.

If you are interested, I would add a new function to your `controller` along the lines of `getTasksFor(<typeOfTask>)` returning a list of tasks matching the type. Then this current function being the view would simply iterate over it to display them.

In other words "model" holds the data, "controller" filters/manipulates the data, and "view" just displays the data.

File: src/modules.js [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+            } else {
+                for (let i = 0; i< tasksObj.tasks.length; i++) {
+                        let currentMonth = new Date().getMonth()
+                        let taskMonth = parseISO(tasksObj.tasks[i].dueDate).getMonth();
+                        if (currentMonth == taskMonth) {
+                            displayTask(tasksObj.tasks[i])
+                        }
+                    }
+                }
+            } else if (list == "priorityLevel") {
```
Comment (akouthoofd) [inline src/modules.js]: Looks like the indentation for here on is accidentally one indent too deep. :P

File: src/modules.js [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+                if (projList[i].name == projName) {
+                    projList.splice(i, 1)
+                }
+            };
+            let taskList = dataModel.allTasks.tasks;
+            for (let i =0; i < taskList.length; i++) {
+                if (taskList[i].list == projName) {
+                    taskList.splice(i, 1)
+                }
+                controller.updateTaskIndex();
```
Comment (akouthoofd) [inline src/modules.js]: Almost! Just not quite MVC 😬 So general rule when it comes to MVC. The View really shouldn't have direct access to the dataModel, though in the cases it does it should never update it. That's the controller's job. This was the right idea; I think some of this was just lost in translation. The controller should really be the one to splice this `taskList` just as it runs the `updateTaskIndex`. I would even say that everything from line 188 to 202 probably should be a function in controller.

File: src/modules.js [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+        "list": "none",
+        "index": index
+        };
+        addTaskToStorage(task);
+        DOM.displayTask(task);
+    };
+
+    const addTaskToStorage = (task) => {
+        dataModel.allTasks.tasks.push(task);
+        JSON.parse(localStorage.getItem('allTasks'));
```
Comment (akouthoofd) [inline src/modules.js]: Unless there is some black magic happening here that I missed I'm pretty sure this parse literally does nothing :P

File: src/modules.js [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+            let main = document.querySelector('#main');
+            main.removeChild(img.parentElement);
+            dataModel.allTasks.tasks.splice(index, 1);
+            updateTaskIndex();
+            updateCardIndex();
+        };
+    };
+
+    const updateTaskIndex = () => {
+        let t = dataModel.allTasks.tasks
```
Comment (akouthoofd) [inline src/modules.js]: I'm not going to say don't use single letter variables for real things buuuuut... 🤣 

Joking aside, this is purely a readability thing. Since this is contained inside a method that's sub 15 lines it's easy enough to follow, but semantic naming (e.g. a name that explain what you want to do with it, like `tasksToReindex`) can go a long way. :)

File: src/modules.js [hunk 18]
```diff
@@ -0,0 +1,450 @@
...
+        for(let i = 0; i < t.length; i++) {
+            t[i].index = i
+        };
+        localStorage.setItem('allTasks', JSON.stringify(dataModel.allTasks))
+    };
+
+    const updateCardIndex = () => {
+        let cards = document.querySelectorAll('.divCard');
+        Array.from(cards).forEach(function (c) {
+            c.setAttribute("data-index", Array.from(cards).indexOf(c));
```
Comment (akouthoofd) [inline src/modules.js]: Getting fancy 😜

Performance concerns are not something you really need to worry about at this point, but I will at least mention that this would technically be better from a performance perspective if you made `Array.from(cards)` a variable which you can reuse. So:
```
const allCards = Array.from(cards);
allCards.forEach(function (c) {
    c.setAttribute("data-index", allCards.indexOf(c));
});
```
If you don't make it a variable it means that for every card you look at you have to recreate a new array of the exact same cards just to run `indexOf` on it. It's just wasted computation. :O

File: src/modules.js [hunk 19]
```diff
@@ -0,0 +1,450 @@
...
+        img.onclick = (e) => {
+            let card = img.parentElement;
+            let index = img.parentElement.getAttribute('data-index');
+            let o = dataModel.allTasks.tasks[index];
+            if (card.style.textDecorationLine == "none") {
+                card.style.textDecorationLine = "line-through";
+                o.strike = "strike";
+            } else {
+                card.style.textDecorationLine = "none";
+                o.strike = "none"
```
Comment (akouthoofd) [inline src/modules.js]: Fun fact: If you are setting two variables equal to the same value this is technically valid syntax:
```
card.style.textDecorationLine = o.strike = "none";
```
Some would argue that it's harder to read and should not be used, but now you know! 😁

File: src/modules.js [hunk 20]
```diff
@@ -0,0 +1,450 @@
...
+                let index = img.parentElement.getAttribute('data-index');
+                let t = dataModel.allTasks.tasks[index]; 
+                let selected = sel.value
+                let proj = dataModel.allTasks.customProjects
+                for (let i=0; i< proj.length; i++) {
+                    if (proj[i].name == selected) {
+                        t.list = selected
+                    }
+                    localStorage.setItem('allTasks', JSON.stringify(dataModel.allTasks))
+                }
```
Comment (akouthoofd) [inline src/modules.js]: If I understand correctly you want to validate that the selected project actual exists before setting it to `t.list`? If that's that case there is a function for that called `contains`. That would make this loop turn into:
```
if (proj.contains(selected)) {
    t.list = selected;
}
localStorage.setItem('allTasks', JSON.stringify(dataModel.allTasks));
```

File: src/modules.js [hunk 21]
```diff
@@ -0,0 +1,450 @@
...
+        };
+        addProjectToStorage(project);
+        DOM.addProjectToList(project);
+    };
+
+    const addProjectToStorage = (project) => {
+        let list = JSON.parse(localStorage.getItem('allTasks')); 
+        if ( list == null) {
+            dataModel.allTasks.customProjects.push(project);
+            localStorage.setItem('allTasks', JSON.stringify(dataModel.allTasks))
```
Comment (akouthoofd) [inline src/modules.js]: I assume these use to do something different at one point, but the logic here is literally the same for both conditions. This whole function could just be these two lines. :P

File: src/modules.js [hunk 22]
```diff
@@ -0,0 +1,450 @@
...
+    const createAddToProject = (img) => {
+        img.onclick = (e) => {
+            let list = JSON.parse(localStorage.getItem('allTasks'));
+            list = list.customProjects
+            if (list.length == 0) {
+                let div = document.createElement('div');
+                div.classList.add('projectAddPopup');
+                div.innerText= `Create a Project first
+                
+                `;
```
Comment (akouthoofd) [inline src/modules.js]: If you were looking for a way to add newlines to something you can use `\n` instead. So this would be:
```
div.innerText= 'Create a Project first\n\n';
```
A lot easier to read compared to whitespace too.

File: src/modules.js [hunk 23]
```diff
@@ -0,0 +1,450 @@
...
+        } else {
+            getTasks(taskStorage);
+            getProjects(taskStorage);
+            localStorage.removeItem('allTasks');
+            localStorage.setItem('allTasks', JSON.stringify(dataModel.allTasks));
+        };
+    };
+
+    const getTasks = (taskStorage) => {
+        taskStorage = taskStorage.tasks
```
Comment (akouthoofd) [inline src/modules.js]: Haha I see what you were going for here. Reducing the number of variables by simple reusing an existing one. This will likely work no issue here, however it's usually bad practice to do this sort of thing. The reasons are complicated, but in the simplest terms overriding an existing variable adds confusions that can easily be avoided. It also loosely falls into the "side-effect" category that I've mentioned previously.

File: src/modules.js [hunk 24]
```diff
@@ -0,0 +1,450 @@
...
+        if (taskStorage == null) {
+            return
+        } else {
+            for (let i = 0; i< taskStorage.length; i++) {
+                dataModel.allTasks.tasks.push(taskStorage[i]) 
+            };
+            DOM.displayDefaultList();
+        };
+    };
```
Comment (akouthoofd) [inline src/modules.js]: This is a nice and small function for me to show a few of the design concepts I've mentioned in other comments:
```
        const newTaskList = taskStorage.tasks;
        if (newTaskList === null) {
            return;
        }
        for (const task of newTaskList) {
            dataModel.allTasks.tasks.push(task); 
        }
        DOM.displayDefaultList();
```
So the new variable is const by default and has a semantic name (explaining what we want from it), the guard-clause is here to exit early preventing nesting, triple equals just by default to avoid possibility of type-coercion, am the use of the "syntactic sugar" for-of loop to abstract away the unneeded index.

All things to either prevent problems before they happen. :)

**Review by akouthoofd — COMMENTED** [review]
Looking great! You're definitely getting the basics down with only minor things here and there which will just be fixed with practice over time. Overall good design and very good format for just getting started 😄 Hopefully you'll find the info here useful, though apologies if it's a bit long winded! Cheers!

File: src/modules.js (hunk 4 above)
Comment (jelewe) [reply src/modules.js]: this is so smart!!! point 2 is so clever!
